src/pages/HabitTracker/Habits.tsx

```diff
@@ -1,87 +1,158 @@
+  HabitWrapper,
+  Header,
+  HeaderText,
+  HeaderButtons,
+  StyledHabitList,
+  DatePickerWrapper,
+} from "../../components/habittracker/habit-components";
+import { HabitList } from "../../components/habittracker/HabitList";
+import { LocalStorage } from "../../common/local-storage";
+import { LogAction } from "../../common/local-storage";
+import {
+  faPlus,
+  faPencilAlt,
+  faTimes,
+} from "@fortawesome/free-solid-svg-icons";
+import { colors } from "../../common/color-utils";
+
+// Utility function to update habit data in state and localStorage
+const updateHabits = (
+  updatedHabits: Habit[],
+  setHabits: (habits: Habit[]) => void
+) => {
+  setHabits(updatedHabits);
+  localStorage.setItem(LocalStorage.HABITS, JSON.stringify(updatedHabits));
+};
+
+// Utility function to update logging data in state and localStorage
+const updateLogging = (
+  updatedLogging: Logging,
+  setLogging: (logging: Logging) => void
+) => {
+  setLogging(updatedLogging);
+  localStorage.setItem(LocalStorage.HABITS_LOG, JSON.stringify(updatedLogging));
+};
+
+// Utility function to reset habit form
+const resetHabitForm = (
+  setNewHabit: (habit: Habit) => void,
+  setOriginalHabit: (habit: Habit) => void
+) => {
+  setNewHabit({ name: "", count: 0 });
+  setOriginalHabit({ name: "", count: 0 });
+};
+
+// Utility function to initialize logging structure
+const initializeLogging = (
+  logging: Logging,
+  habitName: string,
+  year: number,
+  month: string
+): void => {
+  logging[habitName] = logging[habitName] ?? {};
+  logging[habitName][year] = logging[habitName][year] ?? {};
+  logging[habitName][year][month] = logging[habitName][year][month] ?? {};
+};
+
+// Utility function to manage logging data
+const manageLogging = (
+  habitName: string,
+  selectedDate: Date,
+  logging: Logging,
+  setLogging: (logging: Logging) => void,
+  action: LogAction,
+  habits: Habit[]
+) => {
+  const year = selectedDate.getFullYear();
+  const month = selectedDate
+    .toLocaleString("default", { month: "long" })
+    .toLowerCase();
+  const day = selectedDate.getDate();
+
+  const updatedLogging = cloneDeep(logging);
+
+  // Initialize logging structure
+  initializeLogging(updatedLogging, habitName, year, month);
+
+  const currentCount = updatedLogging[habitName]?.[year]?.[month]?.[day] || 0;
+
+  if (action === LogAction.ADD) {
+    // Find the habit to check its count
+    const habit = habits.find((h) => h.name === habitName);
+
+    // Ensure the habit exists and the log count doesn't exceed the habit count
+    if (habit && currentCount < habit.count) {
+      updatedLogging[habitName][year][month][day] = currentCount + 1;
+    }
+  } else if (action === LogAction.REMOVE && currentCount > 0) {
+    updatedLogging[habitName][year][month][day] = currentCount - 1;
+
+    // Remove the day entry if the log count drops to zero
+    if (updatedLogging[habitName][year][month][day] === 0) {
+      delete updatedLogging[habitName][year][month][day];
```

Comment:
you are only deleting the day entry if the log drops to zero. What if there was only one day logged in the month. You should check for that and remove the month entry. Similarly, remove year entry if for that year, there was only one month entry.